objects/src/mock/chain.rs

```diff
@@ -55,3 +55,401 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+            Ok(Self {
+                chain: ChainMmr::default(),
+                data: DefaultMerkleStore::default(),
+                blocks: vec![],
+                nullifiers: TieredSmt::default(),
+                accounts: SimpleSmt::new(64).expect("depth too big for SimpleSmt"),
+                rng,
+                account_id_builder,
+                objects: Objects::new(),
+                pending_objects: Objects::new(),
+            })
+        }
+
+        /// Creates the next block.
+        ///
+        /// This will also make all the objects currently pending available for use.
+        pub fn seal_block(&mut self) -> Result<BlockHeader, MerkleError> {
+            let block_num: u64 = self.blocks.len().try_into().expect("usize to u64 failed");
+            let block_num: Felt = block_num.into();
+
+            for account in self.pending_objects.accounts.iter() {
+                let id: Felt = account.id().into();
+                self.accounts.update_leaf(id.as_int(), account.hash().into())?;
+            }
+
+            // TODO:
+            // - resetting the nullifier tree once defined at the protocol level.
+            // - insering only nullifier from transactions included in the batches, once the batch
+            // kernel has been implemented.
+            for nullifier in self.pending_objects.nullifiers.iter() {
+                self.nullifiers
+                    .insert(*nullifier, [block_num, Felt::ZERO, Felt::ZERO, Felt::ZERO]);
+            }
+            let notes = self.pending_objects.build_notes_tree()?;
+
+            let previous = self.blocks.last();
+            let peaks = self.chain.mmr().accumulator();
+            let chain_root: Digest = peaks.hash_peaks().into();
+            let account_root = self.accounts.root();
+            let prev_hash = previous.map_or(Digest::default(), |header| header.hash());
+            let nullifier_root = self.nullifiers.root();
+            let note_root = notes.root();
+            let version = Felt::ZERO;
+            let timestamp =
+                previous.map_or(TIMESTAMP_START, |header| header.timestamp() + TIMESTAMP_STEP);
+
+            // TODO: Set batch_root and proof_hash to the correct values once the kernel is
+            // available.
+            let batch_root = Digest::default();
+            let proof_hash = Digest::default();
+
+            let header = BlockHeader::new(
+                prev_hash,
+                block_num,
+                chain_root,
+                account_root,
+                nullifier_root,
+                note_root,
+                batch_root,
+                proof_hash,
+                version,
+                timestamp,
+            );
+
+            self.blocks.push(header);
+            self.chain.mmr_mut().add(header.hash());
+            self.objects.update_with(&mut self.pending_objects, header, &notes);
+            self.data.extend(self.accounts.inner_nodes());
```

Comment:
It is a bit unfortunate that we have to iterate over the complete tree on every block. But this should be fine for the this usecase.

Reply:
Yes - I think this is fine for now. At some point later, we'll probably have a separate tree for the account DB (similar to how we handle nullifiers now).